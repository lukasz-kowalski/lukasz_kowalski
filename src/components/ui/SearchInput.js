import React from 'react';

class SearchInput extends React.Component {
  handleChange = event => {
    this.props.handleQueryChange(event);
  }

  render() {
    const { id, label, placeholder } = this.props;
    return (
      <div className="form-group">
        <label htmlFor={id} className="sr-only">{label}</label>
        <input
          type="search"
          className="form-control" 
          id={id}
          placeholder={placeholder} 
          onChange={this.handleChange}
        />
      </div>
    );
  }
}

export default SearchInput;
