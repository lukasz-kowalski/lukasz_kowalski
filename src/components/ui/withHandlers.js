import React from 'react';

function withHandlers(WrappedComponent) {
  return class WithHandlers extends React.Component {
    state = {
      isValid: true,
      value: ''
    }
   
    componentDidUpdate(prevProps) {
      if (prevProps.doValidate !== this.props.doValidate && this.props.doValidate === true) {
        this.validate()
      }
    }
  
    handleChange = async event => {
      const value = event.target.value;
      await this.setState({
        value
      });
      this.validate();
      this.props.handleInputValue(this.props.id, this.state.value);
    }
    
    isEmpty = () => {
      const isValid = this.state.value.length > 0;
      this.setState({
        isValid
      });
      this.props.validateForm(this.props.id, isValid);
    }
  
    validate = () => {
      if (this.props.isRequired) {
        this.isEmpty();
      }
    }
  
    render() {
      const newProps = {
        ...this.props,
        isValid: this.state.isValid,
        value: this.state.value,
        handleChange: this.handleChange,
        validate: this.validate
      }
      return <WrappedComponent {...newProps} />
    }
  }
}

export default withHandlers;
