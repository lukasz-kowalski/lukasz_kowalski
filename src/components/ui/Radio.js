import React from 'react';
import withHandlers from './withHandlers';

const Radio = ({ radioData, id, label, isRequired, errorMsg, handleChange, isValid, reference, value }) => (
  <React.Fragment>
    <label htmlFor={id}>
      {label}{isRequired && <span className="blue">*</span>}
    </label>
    <div className="form-check" id={id}>
      {radioData && radioData.map((item, index) =>
        <React.Fragment key={item.id + index}>
          <input 
            className={`form-check-input ${!isValid ? 'is-invalid' : ''}`}
            type="radio"
            name={id} 
            id={item.id} 
            value={item.value}
            ref={index === 0 ? reference : undefined}  // set focus on first radio button
            checked={value === item.value}
            onChange={handleChange}
          />
          <label className="form-check-label radio-label" htmlFor={item.id}>
            {item.label}
          </label>
        </React.Fragment>
      )}
      <div className="invalid-feedback">
        {errorMsg}
      </div>
    </div>
  </React.Fragment>
);

export default withHandlers(Radio);
