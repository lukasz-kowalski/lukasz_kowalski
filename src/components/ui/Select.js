import React from 'react';
import withHandlers from './withHandlers';

const Select = ({ id, label, options, isRequired, errorMsg, handleChange, value, isValid, reference }) => (
  <div className="form-group">
    <label htmlFor={id}>
      {label}{isRequired && <span className="blue">*</span>}
    </label>
    <select
      className={`form-control ${!isValid ? 'is-invalid' : ''}`}
      id={id} 
      name={id}
      onChange={handleChange}
      ref={reference}
      value={value}
    >
      <option value="">--Please choose an option--</option>
      {options.length > 0 && options.map((item, index) => 
        <option key={'species' + index} value={item}>{item}</option>
      )}
    </select>
    <div className="invalid-feedback">
      {errorMsg}
    </div>
  </div>
);

export default withHandlers(Select);
export { Select };
