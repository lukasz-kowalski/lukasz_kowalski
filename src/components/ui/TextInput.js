import React from 'react';
import withHandlers from './withHandlers';

const TextInput = ({ id, label, isRequired, placeholder, errorMsg, handleChange, validate, isValid, value, reference }) => (
  <div className="form-group">
    <label htmlFor={id}>
      {label}{isRequired && <span className="blue">*</span>}
    </label>
    <input
      type="text"
      className={`form-control ${!isValid ? 'is-invalid' : ''}`}
      id={id}
      name={id} 
      placeholder={placeholder} 
      onChange={handleChange}
      onBlur={validate}
      ref={reference}
      value={value}
    />
    <div className="invalid-feedback">
      {errorMsg}
    </div>
  </div>
);

export default withHandlers(TextInput);
