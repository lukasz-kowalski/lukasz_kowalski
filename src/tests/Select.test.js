import React from 'react';
import { shallow } from 'enzyme';
import { Select } from '../components/ui/Select';

let wrapper;

const options = [
  'Option1',
  'Option2',
  'Option3'
];

const handleChange = jest.fn();

beforeEach(() => {
  wrapper = shallow(
  <Select
    id="character-species"
    label="Choose character species"
    isValid={true}
    options={options}
    handleChange={handleChange}
  />);
})

test('should successful render Select', () => {
  expect(wrapper).toMatchSnapshot();
})

test('should call handleChange function on change', () => {
  const event = { target: { value: 'Test' } };
  wrapper.find('#character-species').simulate('change', event);
  expect(handleChange).toBeCalledWith(event);
})
