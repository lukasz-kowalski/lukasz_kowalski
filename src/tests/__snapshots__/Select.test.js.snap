// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should successful render Select 1`] = `
<div
  className="form-group"
>
  <label
    htmlFor="character-species"
  >
    Choose character species
  </label>
  <select
    className="form-control "
    id="character-species"
    name="character-species"
    onChange={[MockFunction]}
  >
    <option
      value=""
    >
      --Please choose an option--
    </option>
    <option
      key="species0"
      value="Option1"
    >
      Option1
    </option>
    <option
      key="species1"
      value="Option2"
    >
      Option2
    </option>
    <option
      key="species2"
      value="Option3"
    >
      Option3
    </option>
  </select>
  <div
    className="invalid-feedback"
  />
</div>
`;
